### PR TITLE
[ORC] Add signext on @sum() arguments in test.

### LINKED ISSUE #112503: Assertion "Narrow integer argument must have a valid extension type." failed.
@sum seems to have extension attributes missing for the returned value as well as for both of the arguments:

```
Testing: 
FAIL: LLVM-Unit :: ExecutionEngine/Orc/./OrcJITTests/7/21 (1 of 79162)
******************** TEST 'LLVM-Unit :: ExecutionEngine/Orc/./OrcJITTests/7/21' FAILED ********************
Script(shard):
--
GTEST_OUTPUT=json:/home/ijonpan/llvm-project/build/unittests/ExecutionEngine/Orc/./OrcJITTests-LLVM-Unit-344154-7-21.json GTEST_SHUFFLE=0 GTEST_TOTAL_SHARDS=21 GTEST_SHARD_INDEX=7 /home/ijonpan/llvm-
project/build/unittests/ExecutionEngine/Orc/./OrcJITTests
--

Note: This is test shard 8 of 21.
[==========] Running 8 tests from 7 test suites.
[----------] Global test environment set-up.
[----------] 2 tests from CoreAPIsStandardTest
[ RUN      ] CoreAPIsStandardTest.RemoveSymbolsTest
[       OK ] CoreAPIsStandardTest.RemoveSymbolsTest (0 ms)
[ RUN      ] CoreAPIsStandardTest.GeneratorTest
[       OK ] CoreAPIsStandardTest.GeneratorTest (0 ms)
[----------] 2 tests from CoreAPIsStandardTest (0 ms total)

[----------] 1 test from ExecutorAddrTest
[ RUN      ] ExecutorAddrTest.PtrConversion
[       OK ] ExecutorAddrTest.PtrConversion (0 ms)
[----------] 1 test from ExecutorAddrTest (0 ms total)

[----------] 1 test from MapperJITLinkMemoryManagerTest
[ RUN      ] MapperJITLinkMemoryManagerTest.Coalescing
[       OK ] MapperJITLinkMemoryManagerTest.Coalescing (0 ms)
[----------] 1 test from MapperJITLinkMemoryManagerTest (0 ms total)

[----------] 1 test from OrcCAPITestBase
[ RUN      ] OrcCAPITestBase.EnableDebugSupport
ERROR: Missing extension attribute of returned value from function:
i32 @sum(i32, i32)

```
@lhames @nikic

### PR DESCRIPTION
**Pull request by JonPsson1**
Use static std::strings for the textual representations of @sum(). The usage of static variables is already there, e.g. with TargetTriple, so I have assumed that this is thread-safe.

The TargetTriple is not the full Triple class, but just the string representation. If the complete version of this is needed, the full Triple is needed so that it can be passed to TargetLibraryInfo getExtAttrForI32Param/Return. For now, there is just a string recognition of s390 which does need the extensions. Maybe this is ok for now as this is just a test, after all..?

Fixes: #112503

### CODE REVIEW

**Comment by JonPsson1** [comment]
Ping!

File: llvm/unittests/ExecutionEngine/Orc/OrcCAPITest.cpp [hunk 1]
```diff
@@ -90,6 +90,35 @@ class OrcCAPITestBase : public testing::Test {
 
     LLVMOrcDisposeLLJIT(J);
     TargetSupported = true;
+
+    // Create test functions in text format, with the proper extension
+    // attributes.
+    if (SumExample.empty()) {
+      std::string I32RetExt = "";
+      std::string I32ArgExt = "";
```
Comment (lhames) [inline llvm/unittests/ExecutionEngine/Orc/OrcCAPITest.cpp:98]: These can just be `StringRef`s.

File: llvm/unittests/ExecutionEngine/Orc/OrcCAPITest.cpp [hunk 2]
```diff
@@ -90,6 +90,35 @@ class OrcCAPITestBase : public testing::Test {
...
+      if (StringRef(TargetTriple).starts_with("s390x-ibm-linux"))
+        I32RetExt = I32ArgExt = "signext ";
+
+      std::ostringstream OS;
+      OS << "define " << I32RetExt << " i32 "
+         << R"(@sum()" << "i32 " << I32ArgExt << "%x, i32 " << I32ArgExt << "%y)"
```
Comment (lhames) [inline llvm/unittests/ExecutionEngine/Orc/OrcCAPITest.cpp:104]: I think `R"(@sum()"` could just be `"@sum("`.

File: llvm/unittests/ExecutionEngine/Orc/OrcCAPITest.cpp [hunk 3]
```diff
@@ -90,6 +90,35 @@ class OrcCAPITestBase : public testing::Test {
...
     LLVMOrcDisposeLLJIT(J);
     TargetSupported = true;
+
+    // Create test functions in text format, with the proper extension
+    // attributes.
+    if (SumExample.empty()) {
+      std::string I32RetExt = "";
+      std::string I32ArgExt = "";
+      if (StringRef(TargetTriple).starts_with("s390x-ibm-linux"))
+        I32RetExt = I32ArgExt = "signext ";
```
Comment (lhames) [inline llvm/unittests/ExecutionEngine/Orc/OrcCAPITest.cpp:100]: Is there a way that we could move this out into a common header location that could be used elsewhere in LLVM, or at least in the tests?

E.g.
```c++
StringRef getExtension(const Triple &TT) {
  switch (TT.getArch()) {
  case Triple::s390x:
    if (TT.getVendor() == Triple::IBM && TT.getOS() == Triple::Linux)
      return "signext ";
    break;
  default:
    break;
  }
  return "";
}
```

I'm guessing that we'll need this for Kaleidoscope and some other examples and tests too?

Comment (lhames) [reply llvm/unittests/ExecutionEngine/Orc/OrcCAPITest.cpp:100]: Just re-read your comment -- I think a common utility like the above could be handy, but doesn't apply in this C API test where we don't have an `llvm::Triple` available.

File: llvm/unittests/ExecutionEngine/Orc/OrcCAPITest.cpp (hunk 2 above)
Comment (JonPsson1) [reply llvm/unittests/ExecutionEngine/Orc/OrcCAPITest.cpp:104]: yep

**Comment by JonPsson1** [comment]
ping - is the new version like something you were thinking of?

**Review by lhames — APPROVED** [review]
Looks good to me. Thanks @JonPsson1!

**Pull request merged**